apps/emqx_resource/src/emqx_resource.erl

```diff
@@ -418,24 +418,39 @@ query(ResId, Request, Opts) ->
             emqx_resource_buffer_worker:async_query(ResId, Request, Opts)
     end.
 
-get_query_mode_error(ResId, Opts) ->
-    case maps:get(query_mode_cache_override, Opts, true) of
-        false ->
-            case Opts of
-                #{query_mode := QueryMode} ->
-                    {QueryMode, ok};
-                _ ->
-                    {async, unhealthy_target}
-            end;
-        true ->
-            case emqx_resource_manager:lookup_cached(ResId) of
-                {ok, _Group, #{query_mode := QM, error := Error}} ->
-                    {QM, Error};
-                {error, not_found} ->
-                    {error, not_found}
-            end
+get_query_mode_error(_ResId, Request, Opts = #{connector_resource_id := ResId}) ->
+    do_get_query_mode_error(ResId, Request, Opts);
+get_query_mode_error(ResId, Request, Opts) ->
+    do_get_query_mode_error(ResId, Request, Opts).
+
+do_get_query_mode_error(ResId, Request, Opts) ->
+    case emqx_resource_manager:lookup_cached(ResId) of
+        {ok, _Group, ResourceData} ->
+            QM = get_query_mode(ResourceData, Opts),
+            Error = get_error(Request, ResourceData),
+            {QM, Error};
+        {error, not_found} ->
+            {error, not_found}
     end.
 
+get_query_mode(_ResourceData, #{query_mode := QM}) ->
+    QM;
+get_query_mode(#{query_mode := QM}, _Opts) ->
+    QM.
+
+get_error(Request, #{added_channels := #{} = Channels} = ResourceData) when
+    is_tuple(Request), is_binary(element(1, Request))
+->
+    ChannelId = element(1, Request),
```

Comment:
```suggestion
get_error({ChannelId, _Msg} = Request, #{added_channels := Channels} = ResourceData) when
    is_binary(ChannelId)
->
```

Reply:
But I think we can use the requested resource id rather than taking it from the request itself, as it'll be a channel id for actions.